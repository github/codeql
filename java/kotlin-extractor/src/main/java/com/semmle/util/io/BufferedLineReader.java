package com.semmle.util.io;

import java.io.BufferedReader;
import java.io.Closeable;
import java.io.IOException;
import java.io.Reader;

import com.semmle.util.files.FileUtil;

/**
 * A custom buffered reader akin to {@link BufferedReader}, except that it preserves
 * line terminators (and so its {@code readLine()} method is called
 * {@link #readLineAndTerminator()}). The other {@link Reader} methods should not
 * be called, and will throw.
 */
public class BufferedLineReader implements Closeable {
	private final char[] buffer = new char[8192];
	private  int nextChar = 0, nChars = 0;
	private final Reader in;
	
	public BufferedLineReader(Reader in) {
		this.in = in;
	}

	/**
	 * Read the string up to and including the next CRLF or LF terminator. This method
	 * may return a non-terminated string at EOF, or if a line is too long to fit in the
	 * internal buffer. Calls will block until enough data has been read to fill the
	 * buffer or find a line terminator.
	 * @return The next line (or buffer-full) of text.
	 * @throws IOException if the underlying stream throws.
	 */
	public String readLineAndTerminator() throws IOException {
		int terminator = findNextLineTerminator();
		if (terminator == -1)
			return null;
		String result = new String(buffer, nextChar, terminator - nextChar + 1);
		nextChar = terminator + 1;
		return result;
	}
	
	/**
	 * Get the index of the last character that should be included in the next line.
	 * Usually, this is the LF in a LF or CRLF line terminator, but it might be the
	 * end of the buffer (if it is full, and no newlines are present), or it may be
	 * -1 (but only if EOF has been reached, and the buffer is currently empty).
	 * The first character of the line is pointed to by {@link #nextChar}, which
	 * may be modified by this method if the buffer is refilled.
	 */
	private int findNextLineTerminator() throws IOException {
		int alreadyChecked = 0;
		do {
			for (int i = nextChar + alreadyChecked; i < nChars; i++) {
				if (buffer[i] == '\r' && i+1 < nChars && buffer[i+1] == '\n')
					return i+1; // CRLF
				else if (buffer[i] == '\n')
					return i; // LF
			}
			
			// We didn't find a full newline in the existing buffer: Try to fill.
			alreadyChecked = nChars - nextChar;
			int newlyRead = fill();
			if (newlyRead <= 0)
				return nChars - 1;
		} while (true);
	}
	
	/**
	 * Block until at least one character from the underlying stream is read,
	 * or EOF is reached.
	 */
	private int fill() throws IOException {
		if (nextChar >= nChars) {
			// No unread characters.
			nextChar = 0;
			nChars = 0;
		} else if (nextChar > 0) {
			// Some unread characters.
			System.arraycopy(buffer, nextChar, buffer, 0, nChars - nextChar);
			nChars = nChars - nextChar;
			nextChar = 0;
		}
		
		// Is the buffer full?
		if (nChars == buffer.length)
			return 0;
		
		int read;
		do {
			read = in.read(buffer, nChars, buffer.length - nChars);
		} while (read == 0);
		
		if (read > 0) {
			nChars += read;
		}
		return read;
	}

	@Override
	public void close() {
		FileUtil.close(in);
	}
}
