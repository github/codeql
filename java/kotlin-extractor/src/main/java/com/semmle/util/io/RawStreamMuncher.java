package com.semmle.util.io;

import com.semmle.util.exception.Exceptions;
import com.semmle.util.files.FileUtil;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

/**
 * A thread that copies data from an input stream to an output stream. When
 * the input stream runs out, it closes both the input and output streams.
 */
public class RawStreamMuncher extends Thread {
	private final InputStream in;
	private final OutputStream out;

	public RawStreamMuncher(InputStream in, OutputStream out) {
		this.in = in;
		this.out = out;
	}

	@Override
	public void run() {
		try {
			StreamUtil.copy(in, out);
		} catch (IOException e) {
			Exceptions.ignore(e, "When the process exits, a harmless IOException will occur here");
		} finally {
			FileUtil.close(in);
			FileUtil.close(out);
		}
	}
}
