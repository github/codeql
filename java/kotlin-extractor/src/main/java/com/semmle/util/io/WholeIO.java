package com.semmle.util.io;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStreamWriter;
import java.io.UnsupportedEncodingException;
import java.io.Writer;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.nio.file.OpenOption;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.util.Arrays;
import java.util.regex.Pattern;

import com.semmle.util.array.ArrayUtil;
import com.semmle.util.data.IntRef;
import com.semmle.util.exception.ResourceError;
import com.semmle.util.files.FileUtil;

/**
 * A class that allows bulk operations on entire files,
 * reading or writing them as {@link String} values.
 * 
 * This is intended to address the woeful inadequacy of
 * the Java standard libraries in this area.
 */
public class WholeIO {
	private IOException e;
	
	/**
	 * Regular expression {@link Pattern}
	 */
	private final static Pattern rpLineEndingCRLF = Pattern.compile("\r\n");
	
	/**
	 * The default encoding to use for writing, and for reading if no
	 * encoding can be detected.
	 */
	private final String defaultEncoding;
	
	/**
	 * Construct a new {@link WholeIO} instance using ODASA's default
	 * charset ({@code "UTF-8"}) for all input and output (unless a
	 * different encoding is detected for a file being read).
	 */
	public WholeIO() {
		this("UTF-8");
	}
	
	/**
	 * Construct a new {@link WholeIO} instance using the specified
	 * encoding for all input and output (unless a different encoding
	 * is detected for a file being read).
	 * 
	 * @param encoding The encoding name, e.g. {@code "UTF-8"}.
	 */
	public WholeIO(String encoding) {
		defaultEncoding = encoding;
	}
	
	/**
	 * Open the given file for reading, get the entire content
	 * and return it as a {@link String}. Returns <code>null</code>
	 * on error, in which case you can check the getLastException()
	 * method for the exception that occurred.
	 * 
	 * <b>Warning:</b> This method trims the content of the file, removing
	 * leading and trailing whitespace. Do not use it if you care about file
	 * locations being preserved; use 'read' instead.
	 * 
	 * @param file The file to read
	 * @return The <b>trimmed</b> contents of the file, or <code>null</code> on error.
	 */
	public String readAndTrim(File file) {
		e = null;
		FileInputStream f = null;
		try {
			f = new FileInputStream(file);
			String contents = readString(f);
			return contents == null ? null : contents.trim();
		} catch (IOException e) {
			this.e = e;
			return null;
		} finally {
			FileUtil.close(f);
		}
	}

	/**
	 * Open the given filename for writing and dump the given
	 * {@link String} into it. Returns <code>false</code>
	 * on error, in which case you can check the getLastException()
	 * method for the exception that occurred. Tries to create any
	 * enclosing directories that do not exist.
	 * 
	 * @param filename The name of the file to write to
	 * @param contents the string to write out
	 * @return the success state
	 */
	public boolean write(String filename, String contents) {
		return write(new File(filename), contents);
	}

	/**
	 * Open the given filename for writing and dump the given
	 * {@link String} into it. Returns <code>false</code>
	 * on error, in which case you can check the getLastException()
	 * method for the exception that occurred. Tries to create any
	 * enclosing directories that do not exist.
	 * 
	 * @param file The file to write to
	 * @param contents the string to write out
	 * @return the success state
	 */
	public boolean write(File file, String contents) {
		return write(file, contents, false);
	}

	/**
	 * Open the given path for writing and dump the given
	 * {@link String} into it. Returns <code>false</code>
	 * on error, in which case you can check the getLastException()
	 * method for the exception that occurred. Tries to create any
	 * enclosing directories that do not exist.
	 *
	 * @param path The path to write to
	 * @param contents the string to write out
	 * @return the success state
	 */
	public boolean write(Path path, String contents) {
		return write(path, contents, false);
	}

	/**
	 * Open the given filename for writing and dump the given
	 * {@link String} into it. Throws {@link ResourceError}
	 * if we fail.
	 * 
	 * @param file The file to write to
	 * @param contents the string to write out
	 */
	public void strictwrite(File file, String contents) {
		strictwrite(file, contents, false);
	}

	/**
	 * Open the given path for writing and dump the given
	 * {@link String} into it. Throws {@link ResourceError}
	 * if we fail.
	 *
	 * @param path The path to write to
	 * @param contents the string to write out
	 */
	public void strictwrite(Path path, String contents) {
		strictwrite(path, contents, false);
	}

	/**
	 * This is the same as {@link #write(File,String)},
	 * except that this method allows appending to an existing file.
	 * 
	 * @param file the file to write to
	 * @param contents the string to write out
	 * @param append whether or not to append to any existing file
	 * @return the success state
	 */
	public boolean write(File file, String contents, boolean append) {
		if (file.getParentFile() != null)
			file.getParentFile().mkdirs();
		
		FileOutputStream fos = null;
		try {
			fos = new FileOutputStream(file, append);
			Writer writer = new OutputStreamWriter(fos, Charset.forName(defaultEncoding));
			writer.append(contents);
			writer.close();
			return true;
		} catch (IOException e) {
			this.e = e;
			return false;
		} finally {
			FileUtil.close(fos);
		}
	}

	/**
	 * This is the same as {@link #write(Path,String)},
	 * except that this method allows appending to an existing file.
	 *
	 * @param path the path to write to
	 * @param contents the string to write out
	 * @param append whether or not to append to any existing file
	 * @return the success state
	 */
	public boolean write(Path path, String contents, boolean append) {
		try {
			if (path.getParent() != null)
				Files.createDirectories(path.getParent());

			try (Writer writer = Files.newBufferedWriter(path, Charset.forName(defaultEncoding),
					StandardOpenOption.CREATE, StandardOpenOption.WRITE,
					append ? StandardOpenOption.APPEND : StandardOpenOption.TRUNCATE_EXISTING)) {
				writer.append(contents);
			}
		} catch (IOException e) {
			this.e = e;
			return false;
		}
		return true;
	}

	/**
	 * This is the same as {@link #strictwrite(File,String)},
	 * except that this method allows appending to an existing file.
	 */
	public void strictwrite(File file, String contents, boolean append) {
		if (!write(file, contents, append))
			throw new ResourceError("Failed to write file " + file, getLastException());
	}

	/**
	 * This is the same as {@link #strictwrite(Path,String)},
	 * except that this method allows appending to an existing file.
	 */
	public void strictwrite(Path path, String contents, boolean append) {
		if (!write(path, contents, append))
			throw new ResourceError("Failed to write path " + path, getLastException());
	}

	/**
	 * Get the exception that occurred during the last call to
	 * read(), if any. If the last read() call completed normally,
	 * this returns null.
	 * @return The last caught exception, or <code>null</code> if N/A.
	 */
	public IOException getLastException() {
		return e;
	}
	
	public String read(File file) {
		InputStream is = null;
		try {
			is = new FileInputStream(file);
			return readString(is);
		}
		catch (IOException e) {
			this.e = e;
			return null;
		}
		finally {
			FileUtil.close(is);
		}
	}

	public String read(Path path) {
		InputStream is = null;
		try {
			is = Files.newInputStream(path);
			return readString(is);
		}
		catch (IOException e) {
			this.e = e;
			return null;
		}
		finally {
			FileUtil.close(is);
		}
	}

	/**
	 * Read the contents of the given {@link File} as text (line endings are normalised to "\n" in the output).
	 *
	 * @param file	The file to read.
	 * @return		The text contents of the file, if possible, or null if the file cannot be read.
	 */
	public String readText(File file) {
		String result = read(file);
		return result != null ? result.replaceAll("\r\n", "\n") : null;
	}

	/**
	 * Read the contents of the given {@link Path} as text (line endings are normalised to "\n" in the output).
	 *
	 * @param path	The path to read.
	 * @return		The text contents of the path, if possible, or null if the file cannot be read.
	 */
	public String readText(Path path) {
		String result = read(path);
		return result != null ? result.replaceAll("\r\n", "\n") : null;
	}
	
	
	/**
	 * Read the contents of the given {@link File}, throwing a {@link ResourceError}
	 * if we fail.
	 */
	public String strictread(File f) {
		String content = read(f);
		if (content == null)
			throw new ResourceError("Failed to read file " + f, getLastException());
		return content;
	}

	/**
	 * Read the contents of the given {@link Path}, throwing a {@link ResourceError}
	 * if we fail.
	 */
	public String strictread(Path f) {
		String content = read(f);
		if (content == null)
			throw new ResourceError("Failed to read path " + f, getLastException());
		return content;
	}

	/**
	 * Read the contents of the given {@link File} as text (line endings are normalised to "\n" in the output).
	 *
	 * @param file				The file to read.
	 * @return					The text contents of the file, if possible.
	 * @throws ResourceError	If the file cannot be read.
	 */
	public String strictreadText(File file) {
		return rpLineEndingCRLF.matcher(strictread(file)).replaceAll("\n");
	}

	/**
	 * Read the contents of the given {@link Path} as text (line endings are normalised to "\n" in the output).
	 *
	 * @param path				The path to read.
	 * @return					The text contents of the path, if possible.
	 * @throws ResourceError	If the path cannot be read.
	 */
	public String strictreadText(Path path) {
		return rpLineEndingCRLF.matcher(strictread(path)).replaceAll("\n");
	}

	/**
	 * Get the entire content of an {@link InputStream}
	 * and interpret it as a {@link String} trying to detect its character set.
	 * Returns <code>null</code> on error, in which case you can check
	 * the getLastException() method for the exception that occurred.
	 * 
	 * @param stream the stream to read from
	 * @return The contents of the file, or <code>null</code> on error.
	 */
	public String readString(InputStream stream) {
		IntRef length = new IntRef(0);
		byte[] bytes = readBinary(stream, length);
		
		if (bytes == null) return null;
		
		try {
			IntRef start = new IntRef(0);
			String charset = determineCharset(bytes, length.get(), start);
			return new String(bytes, start.get(), length.get() - start.get(), charset);
		} catch (UnsupportedEncodingException e) {
			this.e = e;
			return null;
		}
	}

	/**
	 * Get the entire content of an {@link InputStream}
	 * and interpret it as a {@link String} trying to detect its character set.
	 * Throws a {@link ResourceError} on error.
	 *
	 * @param stream the stream to read from
	 * @return the contents of the input stream
	 */
	public String strictReadString(InputStream stream) {
		String content = readString(stream);
		if (content == null)
			throw new ResourceError("Could not read from stream", getLastException());
		return content;
	}

	/**
	 * Get the entire content of an {@link InputStream}, interpreting it
	 * as a sequence of bytes. This removes restrictions regarding invalid
	 * code points that would potentially prevent reading a file's contents
	 * as a String.
	 * 
	 * This method returns <code>null</code> on error, in which case you can
	 * check {@link #getLastException()} for the exception that occurred.
	 * 
	 * @param stream the stream to read from
	 * @return The binary contents of the file, or <code>null</code> on error.
	 */
	public byte[] readBinary(InputStream stream) {
		IntRef length = new IntRef(0);
		byte[] bytes = readBinary(stream, length);
		return bytes == null ? null : Arrays.copyOf(bytes, length.get());
	}

	/**
	 * Get the entire content of an {@link InputStream}, interpreting it
	 * as a sequence of bytes. This removes restrictions regarding invalid
	 * code points that would potentially prevent reading a file's contents
	 * as a String.
	 *
	 * @param stream the stream to read from
	 * @return The binary contents of the file -- always non-null.
	 * @throws ResourceError if an exception occurs during IO.
	 */
	public byte[] strictReadBinary(InputStream stream) {
		byte[] result = readBinary(stream);
		if (result == null)
			throw new ResourceError("Couldn't read from stream", e);
		return result;
	}
	
	/**
	 * Get the entire binary contents of a {@link File} as a sequence of bytes.
	 * 
	 * @param file the file to read
	 * @return the file's contents as a byte[] -- always non-null.
	 * @throws ResourceError if an exception occurs during IO.
	 */
	public byte[] strictReadBinary(File file) {
		FileInputStream stream = null;
		try {
			stream = new FileInputStream(file);
			byte[] result = readBinary(stream);
			if (result == null)
				throw new ResourceError("Couldn't read from file " + file + ".", e);
			return result;
		} catch (FileNotFoundException e) {
			throw new ResourceError("Couldn't read from file " + file + ".", e);
		} finally {
			FileUtil.close(stream);
		}
	}

	/**
	 * Get the entire binary contents of a {@link Path} as a sequence of bytes.
	 * 
	 * @param path the path to read
	 * @return the file's contents as a byte[] -- always non-null.
	 * @throws ResourceError if an exception occurs during IO.
	 */
	public byte[] strictReadBinary(Path path) {
		InputStream stream = null;
		try {
			stream = Files.newInputStream(path);
			byte[] result = readBinary(stream);
			if (result == null)
				throw new ResourceError("Couldn't read from path " + path + ".", e);
			return result;
		} catch (IOException e) {
			throw new ResourceError("Couldn't read from path " + path + ".", e);
		} finally {
			FileUtil.close(stream);
		}
	}

	/**
	 * Get the entire binary contents of a {@link Path} as a sequence of bytes.
	 * 
	 * @param path the path to read
	 * @return the file's contents as a byte[] -- always non-null.
	 */
	public byte[] readBinary(Path path) throws IOException {
		InputStream stream = null;
		try {
			stream = Files.newInputStream(path);
			byte[] result = readBinary(stream);
			if (result == null)
				throw new ResourceError("Couldn't read from path " + path + ".", e);
			return result;
		} finally {
			FileUtil.close(stream);
		}
	}
	
	private byte[] readBinary(InputStream stream, IntRef offsetHolder) {
		try {
			byte[] bytes = new byte[16384];
			int offset = 0;
			int readThisTime;
			do {
				readThisTime = stream.read(bytes, offset, bytes.length - offset);
				if (readThisTime > 0) {
					offset += readThisTime;
					if (offset == bytes.length)
						bytes = safeArrayDouble(bytes);
				}
			} while (readThisTime > 0);
			offsetHolder.set(offset);
			return bytes;
		} catch (IOException e) {
			this.e = e;
			return null;
		}
	}

	/**
	 * Safely attempt to double the length of an array.
	 * @param array The array which want to be doubled
	 * @return a new array that is longer than array 
	 */
	private byte[] safeArrayDouble(byte[] array) {
		if (array.length  >= ArrayUtil.MAX_ARRAY_LENGTH) {
			throw new ResourceError("Cannot stream into array as it exceed the maximum array size");
		}
		// Compute desired capacity
		long newCapacity = array.length * 2L;
		// Ensure it is at least as large as minCapacity
		if (newCapacity < 16)
			newCapacity = 16;
		// Ensure it is at most MAX_ARRAY_LENGTH
		if (newCapacity > ArrayUtil.MAX_ARRAY_LENGTH) {
			newCapacity = ArrayUtil.MAX_ARRAY_LENGTH;
		}
		return Arrays.copyOf(array, (int)newCapacity);
	}

	/**
	 * Try to determine the encoding of a byte[] using a byte-order mark (if present)
	 * Defaults to UTF-8 if none found.
	 */
	private String determineCharset(byte[] bom, int length, IntRef start) {
		start.set(0);
		String ret = defaultEncoding;
		if(length < 2)
			return ret;
		if (length >= 3 && byteToInt(bom[0]) == 0xEF && byteToInt(bom[1]) == 0xBB && byteToInt(bom[2]) == 0xBF) {
			ret = "UTF-8";
			start.set(3);
		} else if (byteToInt(bom[0]) == 0xFE && byteToInt(bom[1]) == 0xFF) {
			ret = "UTF-16BE";
			start.set(2);
		} else if (byteToInt(bom[0]) == 0xFF && byteToInt(bom[1]) == 0xFE) {
			ret = "UTF-16LE";
			start.set(2);
		}
		return ret;
	}
	
	private static int byteToInt(byte b) {
		return b & 0xFF;
	}

}
