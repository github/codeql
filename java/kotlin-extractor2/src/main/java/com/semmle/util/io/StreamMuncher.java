package com.semmle.util.io;

import com.semmle.util.exception.Exceptions;
import com.semmle.util.files.FileUtil;
import com.semmle.util.io.BufferedLineReader;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.PrintStream;

/**
 * A thread that forwards data from one stream to another. It waits for
 * entire lines of input from one stream before writing data to the next
 * stream, and it flushes as it goes.
 */
public class StreamMuncher extends Thread {
	private final InputStream is;
	private PrintStream output;
	private BufferedLineReader reader;

	public StreamMuncher(InputStream is, OutputStream output) {
		this.is = is;
		if (output != null)
			this.output = new PrintStream(output);
	}

	@Override
	public void run() {
		InputStreamReader isr = null;
		try {
			isr = new InputStreamReader(is);
			reader = new BufferedLineReader(isr);
			String line;
			while ((line = reader.readLineAndTerminator()) != null) {
				if (output != null) {
					output.print(line);
					output.flush();
				}
			}
		} catch (IOException e) {
			Exceptions.ignore(e, "When the process exits, a harmless IOException will occur here");
		} finally {
			FileUtil.close(reader);
			FileUtil.close(isr);
		}
	}
}
