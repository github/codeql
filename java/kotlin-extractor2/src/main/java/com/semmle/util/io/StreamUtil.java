package com.semmle.util.io;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.io.Reader;
import java.io.Writer;
import java.nio.charset.StandardCharsets;

import com.semmle.util.exception.CatastrophicError;

/**
 * Utility methods concerning {@link InputStream}s and {@link OutputStream}s.
 */
public class StreamUtil
{
	/**
	 * Copy all bytes that can be read from an {@link InputStream}, into an {@link OutputStream}.
	 *
	 * @param inputStream The InputStream from which to read, until an
	 *          {@link InputStream#read(byte[])} operation returns indicating that the input stream
	 *          has reached its end.
	 * @param outputStream The OutputStream to which all bytes read from {@code inputStream} should be
	 *          written.
	 * @return The number of bytes copied.
	 * @throws IOException from {@link InputStream#read(byte[])} or
	 *           {@link OutputStream#write(byte[], int, int)}
	 * @throws CatastrophicError if either of the streams is {@code null}
	 */
	public static long copy(InputStream inputStream, OutputStream outputStream) throws IOException
	{
		nullCheck(inputStream, outputStream);

		// Copy byte data
		long   total = 0;
		byte[] bytes = new byte[1024];
		int    read;
		while ((read = inputStream.read(bytes)) > 0) {
			outputStream.write(bytes, 0, read);
			total += read;
		}
		return total;
	}
	
	/**
	 * Copy all chars that can be read from a {@link Reader}, into a {@link Writer}.
	 *
	 * @param reader The Reader from which to read, until a {@link Reader#read(char[])} operation
	 *            returns indicating that the reader has reached its end.
	 * @param writer The Writer to which all characters read from {@code reader} should be written.
	 * @return The number of bytes copied.
	 * @throws IOException from {@link Reader#read(char[])} or
	 *             {@link Writer#write(char[], int, int)}
	 * @throws CatastrophicError if either of the streams is {@code null}
	 */
	public static long copy(Reader reader, Writer writer) throws IOException
	{
		nullCheck(reader, writer);

		// Copy byte data
		long   total = 0;
		char[] chars = new char[1024];
		int    read;
		while ((read = reader.read(chars)) > 0) {
			writer.write(chars, 0, read);
			total += read;
		}
		return total;
	}

	/**
	 * Copy at most {@code length} bytes from an {@link InputStream}, into an {@link OutputStream}.
	 * <p>
	 * Note that this method will busy-wait during periods for which the {@code inputStream} cannot
	 * supply any data, but has not reached its end.
	 * </p>
	 *
	 * @param inputStream The InputStream from which to read, until {@code length} bytes have
	 *          been read or {@link InputStream#read(byte[], int, int)} operation returns
	 *          indicating that the input stream has reached its end.
	 * @param outputStream The OutputStream to which all bytes read from {@code inputStream} should be
	 *          written.
	 * @param length The maximum number of bytes to copy
	 * @return The number of bytes copied.
	 * @throws IOException from {@link InputStream#read(byte[], int, int)} or
	 *           {@link OutputStream#write(byte[], int, int)}
	 * @throws CatastrophicError if either of the streams is {@code null}
	 */
	public static long limitedCopy(InputStream inputStream, OutputStream outputStream, long length) throws IOException
	{
		nullCheck(inputStream, outputStream);

		// Copy byte data
		long   total = 0;
		byte[] bytes = new byte[1024];
		int    read;
		while ((read = inputStream.read(bytes, 0, (int) Math.min(bytes.length, length))) > 0) {
			outputStream.write(bytes, 0, read);
			length -= read;
			total += read;
		}
		return total;
	}

	private static void nullCheck(Object input, Object output) {
		CatastrophicError.throwIfAnyNull(input, output);
	}

	/**
	 * Skips over and discards n bytes of data from an input stream. If n is negative then no bytes are skipped.
	 * @param stream the InputStream
	 * @param n the number of bytes to be skipped.
	 * @return false if the end-of-file was reached before successfully skipping n bytes
	 */
	public static boolean skip(InputStream stream, long n) throws IOException {
		if (n <= 0)
			return true;
		long toSkip = n - 1;

		while (toSkip > 0) {
			long skipped = stream.skip(toSkip);
			if (skipped == 0) {
				if(stream.read() == -1)
					return false;
				else
					skipped++;
			}
			toSkip -= skipped;
		}
		if(stream.read() == -1)
			return false;
		else
			return true;
	}

	/**
	 * Reads n bytes from the input stream and returns them. This method will block
	 * until all n bytes are available. If the end of the stream is reached before n bytes are
	 * read it returns just the read bytes.
	 *
	 * @param stream the InputStream
	 * @param n the number of bytes to read
	 * @return the read bytes
	 * @throws IOException if an IOException occurs when accessing the stream
	 * @throws IllegalArgumentException if n is negative
	 */
	public static byte[] readN(InputStream stream, int n) throws IOException {
		if (n < 0) throw new IllegalArgumentException("n must be positive");

		ByteArrayOutputStream bOut = new ByteArrayOutputStream();
		limitedCopy(stream, bOut, n);
		return bOut.toByteArray();
	}

	/**
	 * Reads bytes from the input stream into the given buffer. This method will block
	 * until all bytes are available. If the end of the stream is reached before enough bytes are
	 * read it reads as much as it can.
	 *
	 * @param stream the InputStream
	 * @param buf the buffer to read into
	 * @param offset the offset to read into
	 * @param length the number of bytes to read
	 * @return the total number of read bytes
	 * @throws IOException if an IOException occurs when accessing the stream
	 * @throws IllegalArgumentException if n is negative
	 */
	public static int read(InputStream stream, byte[] buf, int offset, int length) throws IOException {
		if (length < 0) throw new IllegalArgumentException("length must be positive");

		// Copy byte data
		int total = 0;
		int read;
		while ((read = stream.read(buf, offset, length)) > 0) {
			length -= read;
			total += read;
		}

		return total;
	}

	/**
	 * Convenience method for constructing a buffered reader with a UTF8 charset.
	 */
	public static BufferedReader newUTF8BufferedReader(InputStream inputStream) {
		return new BufferedReader(new InputStreamReader(inputStream, StandardCharsets.UTF_8));
	}

	/**
	 * Convenience method for constructing a buffered writer with a UTF8 charset.
	 */
	public static BufferedWriter newUTF8BufferedWriter(OutputStream outputStream) {
		return new BufferedWriter(new OutputStreamWriter(outputStream, StandardCharsets.UTF_8));
	}

}
