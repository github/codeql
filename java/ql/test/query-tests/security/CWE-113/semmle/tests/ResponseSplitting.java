// Test case for
// CWE-113: Improper Neutralization of CRLF Sequences in HTTP Headers ('HTTP Response Splitting')
// http://cwe.mitre.org/data/definitions/113.html

package test.cwe113.cwe.examples;



import java.io.IOException;
import javax.servlet.ServletException;
import javax.servlet.http.Cookie;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

public class ResponseSplitting extends HttpServlet {
	protected void doGet(HttpServletRequest request, HttpServletResponse response)
	throws ServletException, IOException {
		// BAD: setting a cookie with an unvalidated parameter
		// can lead to HTTP splitting
		{
			Cookie cookie = new Cookie("name", request.getParameter("name"));
			response.addCookie(cookie);
		}

		// BAD: setting a header with an unvalidated parameter
		// can lead to HTTP splitting
		response.addHeader("Content-type", request.getParameter("contentType"));
		response.setHeader("Content-type", request.getParameter("contentType"));

		// GOOD: remove special characters before putting them in the header
		{
			String name = removeSpecial(request.getParameter("name"));
			Cookie cookie = new Cookie("name", name);
			response.addCookie(cookie);
		}

		// GOOD: Splicing headers into other headers cannot cause splitting
		response.setHeader("Access-Control-Allow-Origin", request.getHeader("Origin"));
	}

	private static String removeSpecial(String str) {
		return str.replaceAll("[^a-zA-Z ]", "");
	}

	public void addCookieName(HttpServletResponse response, Cookie cookie) {
		// GOOD: cookie.getName() cannot lead to HTTP splitting
		Cookie cookie2 = new Cookie("name", cookie.getName());
		response.addCookie(cookie2);
	}

	public void sanitizerTests(HttpServletRequest request, HttpServletResponse response){
		String t = request.getParameter("contentType");

		// GOOD: whitelist-based sanitization
		response.setHeader("h", t.replaceAll("[^a-zA-Z]", ""));

		// BAD: not replacing all problematic characters
		response.setHeader("h", t.replaceFirst("[^a-zA-Z]", ""));

		// GOOD: replace all line breaks
		response.setHeader("h", t.replace('\n', ' ').replace('\r', ' '));

		// FALSE NEGATIVE: replace only some line breaks
		response.setHeader("h", t.replace('\n', ' '));

		// FALSE NEGATIVE: replace only some line breaks
		response.setHeader("h", t.replaceAll("\r", ""));

		// GOOD: replace all linebreaks with a simple regex
		response.setHeader("h", t.replaceAll("\n", "").replaceAll("\r", ""));

		// GOOD: replace all linebreaks with a complex regex
		response.setHeader("h", t.replaceAll("[\n\r]", ""));

		// GOOD: replace all linebreaks with a complex regex
		response.setHeader("h", t.replaceAll("something|[a\nb\rc]+|somethingelse", ""));
	}
}
