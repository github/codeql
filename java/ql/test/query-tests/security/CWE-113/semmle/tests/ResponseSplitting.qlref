Security/CWE/CWE-113/ResponseSplitting.ql
