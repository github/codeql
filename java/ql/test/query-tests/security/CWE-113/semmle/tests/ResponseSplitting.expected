edges
| ResponseSplitting.java:22:20:22:67 | new Cookie(...) : Cookie | ResponseSplitting.java:23:23:23:28 | cookie |
| ResponseSplitting.java:22:39:22:66 | getParameter(...) : String | ResponseSplitting.java:22:20:22:67 | new Cookie(...) : Cookie |
| ResponseSplitting.java:53:14:53:48 | getParameter(...) : String | ResponseSplitting.java:59:27:59:27 | t : String |
| ResponseSplitting.java:59:27:59:27 | t : String | ResponseSplitting.java:59:27:59:57 | replaceFirst(...) |
nodes
| ResponseSplitting.java:22:20:22:67 | new Cookie(...) : Cookie | semmle.label | new Cookie(...) : Cookie |
| ResponseSplitting.java:22:39:22:66 | getParameter(...) : String | semmle.label | getParameter(...) : String |
| ResponseSplitting.java:23:23:23:28 | cookie | semmle.label | cookie |
| ResponseSplitting.java:28:38:28:72 | getParameter(...) | semmle.label | getParameter(...) |
| ResponseSplitting.java:29:38:29:72 | getParameter(...) | semmle.label | getParameter(...) |
| ResponseSplitting.java:53:14:53:48 | getParameter(...) : String | semmle.label | getParameter(...) : String |
| ResponseSplitting.java:59:27:59:27 | t : String | semmle.label | t : String |
| ResponseSplitting.java:59:27:59:57 | replaceFirst(...) | semmle.label | replaceFirst(...) |
subpaths
#select
| ResponseSplitting.java:23:23:23:28 | cookie | ResponseSplitting.java:22:39:22:66 | getParameter(...) : String | ResponseSplitting.java:23:23:23:28 | cookie | This header depends on a $@, which may cause a response-splitting vulnerability. | ResponseSplitting.java:22:39:22:66 | getParameter(...) | user-provided value |
| ResponseSplitting.java:28:38:28:72 | getParameter(...) | ResponseSplitting.java:28:38:28:72 | getParameter(...) | ResponseSplitting.java:28:38:28:72 | getParameter(...) | This header depends on a $@, which may cause a response-splitting vulnerability. | ResponseSplitting.java:28:38:28:72 | getParameter(...) | user-provided value |
| ResponseSplitting.java:29:38:29:72 | getParameter(...) | ResponseSplitting.java:29:38:29:72 | getParameter(...) | ResponseSplitting.java:29:38:29:72 | getParameter(...) | This header depends on a $@, which may cause a response-splitting vulnerability. | ResponseSplitting.java:29:38:29:72 | getParameter(...) | user-provided value |
| ResponseSplitting.java:59:27:59:57 | replaceFirst(...) | ResponseSplitting.java:53:14:53:48 | getParameter(...) : String | ResponseSplitting.java:59:27:59:57 | replaceFirst(...) | This header depends on a $@, which may cause a response-splitting vulnerability. | ResponseSplitting.java:53:14:53:48 | getParameter(...) | user-provided value |
